### ISSUE TITLE
[napari-workflows] Pyramid levels

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-tasks-core#139]
Branching from https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/31#issuecomment-1280852826

> Regarding pyramid levels: I'd say the user chooses a pyramid level to run the whole workflow in, one can't easily change that within the workflow. Do you still see issues then and what would they be?

My understanding is that:
* Labeling should take place at a user-defined level (sometimes it could be too expensive to do it at level 0)
* Measurement should always take place at level 0, with label arrays upscaled to match the high-res image arrays.

This is easy to achieve if the labels are not produced by napari-worfklows, since we can just upscale them at the very beginning (before running a napari-workflow). But some other napari-workflows involve both labeling and measurement, and thus we cannot select different levels easily.

The simplest workarounds are a bit cumbersome, but we can explore them:
1. We provide a coarse image array as an input, we let the whole workflow run, and at the end we need to rescale the measurement by the appropriate factor. Annoying part: this requires scanning all columns of the dataframe, and "knowing" which ones are lengths/areas/volumes.
2. We construct a custom workflow which takes two images: one at level X (to be used in labeling), the other one at level 0 (to be used for measurement, after labels are upscaled within the workflow). Annoying part: we cannot use a given workflow "as is".

**Comment by jluethi** [comment]
Hmm, that's a fair point. Though I think there are 2 workarounds for this that avoid us having to dive into this:
1) Run 2 separate workflows. The first to get segmentation e.g. at level 2. And then a second workflow that runs at level 0.
2) Have a workflow that handles the resolutions within the workflow, e.g. the workflow downsamples images to create segmentation, upsamples the result again before saving. Some deep-learning libraries we may use can do this built-in I think.

Switching resolutions within a workflow would then be option 3 and can be in the backlog for now. If we want to tackle it, I suggest we go with a version of your option 2. i.e. the user provides 2 different inputs: 1) the image at low res, 2) the same image at high res => would require that we don't just specify the channel, but also the resolution level for inputs and outputs.  I think that can be done as part of the actual workflow by picking 2 different inputs in the workflow as well.

But I think to start with, running workflows at level 0 should be a good start. Once we want to generalize, adding the option to choose the resolution level for inputs should work

**Comment by jluethi** [comment]
Though thinking about my proposed solution 1 now: That also requires us to be able to specify the resolution level we want to process. So that probably needs to be added to the workflow wrapping, as part of the input we specify? And that should default to resolution level 0, but if we want to support processing at lower resolution levels, then an optional `level` parameter would also need to be part of the input & output dictionary.

**Comment by tcompa** [comment]
[I'm splitting this comment into parts, for clarity]

I started to work on this, but then realized that the plan is not clear yet. To make it explicit (but I'm open to other advice): I am now trying to implement a version of @jluethi's option 1:
> Run 2 separate workflows. The first to get segmentation e.g. at level 2. And then a second workflow that runs at level 0 to make measurements.
> That also requires us to be able to specify the resolution level we want to process

This idea of splitting a complex workflow into two different ones is nice in that it provides a "segmentation" fractal task (this could be through cellpose o napari-workflows), and then a "measurement" fractal task (always through napari-workflows). The segmentation one would possibly run at `level>0`, and the measurement one would simply run at level=0` (upscaling any input label array to make it match the image array).

A note:
* Since I understand that napari-workflows could offer much more than just labeling/measurement, are there already some interesting example that we think should be covered in the future? Listing relevant examples would be helpful for the wrapper design.

To-do for this:
- [ ] Add a `level` parameter to the task (but not to each entry in the `input_specs` dictionary).
- [ ] We should include at least a soft check, that raises a warning if `level>0` but some task include `regionprops` (are there other known measurements functions? should this be a hard check?)

**Comment by tcompa** [comment]
A different option is to set a `level` value for each entry of `input_specs`.

TL;DR: I think handling all cases would quickly become a bit complex

Let us look at some examples of `input_specs` dictionaries:
```
input_specs_1 = {
    "input": {"type": "image", "channel": "A01_C01"},
}
input_specs_2 = {
    "slice_img": {"type": "image", "channel": "A01_C01"},
    "slice_img_c2": {"type": "image", "channel": "A01_C01"},
}
input_specs_3 = {
    "dapi_img": {"type": "image", "channel": "A01_C01"},
    "dapi_label_img": {"type": "label", "label_name": "label_DAPI"},
}
```

1. For case 1, it would be very simple to replace the `input` value with `{"type": "image", "channel": "A01_C01", "level": 1}`. Since the corresponding workflow performs some labeling, we would then know that any output label should inherit the `level` parameter from `input`. Note that there is a single input here.
2. For case 2, we already have to enforce an additional check. If we add a `level` to each of the two inputs, the two levels must be equal. This is required later by the workflow, but fractal cannot inspect the workflow in advance to know that. Easiest (but not cleanest) solution: let any input go through, and then fail via `napari-workflows`. Slightly better solution (in my opinion): we force all input levels to be the same (but this brings us back to a global `level` parameter).
3. In this case, in principle, levels are already handled correctly: we take level 0 of both `dapi_img` and `dapi_label_img`, and we upscale the label array to match the image array.

**Comment by tcompa** [comment]
Perhaps this question will help us move forward:
**Which kind of processing will need to take place at arbitrary resolution?**

At the moment, this is only relevant for labeling. Are there other known cases?
If not, the simplest approach would be the one of asking the user to separate the labeling workflow from the rest: we would expose a task-wise `level` parameter, but allow a value `level>0` only for a workflow that satisfies some constraints (e.g. it includes a labeling task, and it does not include any measurement).

**Comment by tcompa** [comment]
This is now in-place, with a code like
```python
    # Characterization of workflow
    input_types = [params["type"] for (name, params) in input_specs.items()]
    output_types = [params["type"] for (name, params) in output_specs.items()]
    are_inputs_all_images = set(input_types) == {"image"}
    are_outputs_all_labels = set(output_types) == {"label"}
    is_labeling_workflow = are_inputs_all_images and are_outputs_all_labels
    if level > 0 and not is_labeling_workflow:
        msg = ( 
            f"{level=}>0 is currently only accepted for labeling workflows, "
            "i.e. those going from image(s) to label(s)"
        ) 
        raise NotImplementedError(msg)
```

TO-DO:
- [ ] Design a new example where the labeling+measurement workflow (e.g. `wf_3.yaml`) is split into two workflows, and labeling is performed at a higher level.

**Comment by tcompa** [comment]
Also to consider: #152

**Comment by jluethi** [comment]
Sorry for only catching up on this now.

Going with option 1 of splitting workflows looks good to me and great that we have a way of passing levels now. Are levels a required argument or do they default to level 0? I'd think defaulting to level 0 makes sense.

I think in general, it's hard to know what operations make sense at what level. A reasonable default assumption is that we don't want to do measurement operations at levels > 0, so your implementation makes sense as a first start.

I'll open an issue to discuss this further whether this assumption should be relaxed and e.g. just be a warning. 

Also, at the moment, there is a global level for the workflow, right? Seems reasonable, I'll need to think whether we'd ever need to relax that.

**Comment by jluethi** [comment]
Ok, I thought it over in detail in https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/159

My conclusion: The current implementation is good. There is a few areas we may want to adapt in the future, but I wouldn't focus on any of them now :)
